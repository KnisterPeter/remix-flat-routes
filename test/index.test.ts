import { defineRoutes } from '@remix-run/dev/config/routes'
import type { RouteManifest } from '../src/index'
import flatRoutes from '../src/index'

describe('define routes', () => {
  it('should define routes for flat-files', () => {
    const flatFiles = [
      '$lang.$ref.tsx',
      '$lang.$ref._index.tsx',
      '$lang.$ref.$.tsx',
      '_index.tsx',
      'healthcheck.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFiles),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should define routes for flat-folders', () => {
    const flatFolders = [
      '$lang.$ref/_index.tsx',
      '$lang.$ref._index/_index.tsx',
      '$lang.$ref.$/_index.tsx',
      '_index/_index.tsx',
      'healthcheck/_index.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFolders),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should define routes for flat-folders on Windows', () => {
    const flatFolders = [
      '$lang.$ref\\_index.tsx',
      '$lang.$ref._index\\_index.tsx',
      '$lang.$ref.$\\_index.tsx',
      '_index\\_index.tsx',
      'healthcheck\\_index.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFolders),
    })
    expect(routes['routes/$lang.$ref._index/_index'].parentId).toBe(
      'routes/$lang.$ref/_index',
    )
    expect(routes).toMatchSnapshot()
  })
  it('should define routes for complex structure', () => {
    const routeList = [
      '_auth.forgot-password.tsx',
      '_auth.login.tsx',
      '_auth.reset-password.tsx',
      '_auth.signup.tsx',
      '_auth.tsx',
      '_landing.about.tsx',
      '_landing.index.tsx',
      '_landing.tsx',
      'app.calendar.$day.tsx',
      'app.calendar.index.tsx',
      'app.calendar.tsx',
      'app.projects.$id.tsx',
      'app.projects.tsx',
      'app.tsx',
      'app_.projects.$id.roadmap.tsx',
      'app_.projects.$id.roadmap[.pdf].tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(routeList),
    })
    expect(routes).toMatchSnapshot()
  })

  it('should correctly nest routes', () => {
    const routeList = [
      'app.$organizationSlug.tsx',
      'app.$organizationSlug.edit.tsx',
      'app.$organizationSlug.projects.tsx',
      'app.$organizationSlug.projects.$projectId.tsx',
      'app.$organizationSlug.projects.$projectId.edit.tsx',
      'app.$organizationSlug.projects.new.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(routeList),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should allow routes to specify different parent routes', () => {
    const routeList = [
      'parent.tsx',
      'parent.some.nested.tsx',
      'parent.some_.nested.route.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(routeList),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should handle params with trailing underscore', () => {
    const routeList = [
      'app.$organizationSlug_._projects.tsx',
      'app.$organizationSlug_._projects.projects.new.tsx',
      'app.$organizationSlug_._projects.projects.$projectId.tsx',
      'app.$organizationSlug_._projects.projects.$projectId.edit.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(routeList),
    })
    expect(routes).toMatchSnapshot()
  })

  it('should ignore non-route files in flat-folders', () => {
    const flatFolders = [
      '$lang.$ref/_layout.tsx',
      '$lang.$ref/component.tsx',
      '$lang.$ref._index/_index.tsx',
      '$lang.$ref._index/style.css',
      '$lang.$ref.$/model.server.ts',
      '_index/_index.tsx',
      'healthcheck/_route.tsx',
    ]
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFolders),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should support markdown routes as flat-files', () => {
    const flatFiles = ['docs.tsx', 'docs.readme.md']
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFiles),
    })
    expect(routes).toMatchSnapshot()
  })
  it('should support markdown routes as flat-folders', () => {
    const flatFolders = ['docs/_layout.tsx', 'docs/readme.route.mdx']
    const routes = flatRoutes('routes', defineRoutes, {
      visitFiles: visitFilesFromArray(flatFolders),
    })
    expect(routes).toMatchSnapshot()
  })
})

function visitFilesFromArray(files: string[]) {
  return (_dir: string, visitor: (file: string) => void, _baseDir?: string) => {
    files.forEach(file => {
      visitor(file)
    })
  }
}

type RouteMapInfo = {
  id: string
  path: string
  file: string
  index?: boolean
  children: string[]
}
function dumpRoutes(routes: RouteManifest) {
  const routeMap = new Map<string, RouteMapInfo>()
  const rootRoute: RouteMapInfo = {
    id: 'root',
    path: '',
    file: 'root.tsx',
    index: false,
    children: [],
  }
  routeMap.set('root', rootRoute)
  Object.entries(routes).forEach(([name, route]) => {
    if (!route.parentId) return
    const parent = routeMap.get(route.parentId)
    if (parent) {
      parent.children.push(name)
    }
    routeMap.set(name, {
      id: name,
      path: route.path,
      file: route.file,
      index: route.index,
      children: [],
    })
  })
  const dump = (route: RouteMapInfo, indent: string) => {
    const getPath = (path?: string) => (path ? `path="${path}" ` : '')
    const getIndex = (index?: boolean) => (index ? 'index ' : '')
    output += `${indent}<Route ${getIndex(route.index)}${getPath(
      route.path,
    )}file="${route.file}"${route.children ? '' : ' /'}>\n`
    if (route.children.length) {
      route.children.forEach((childId: string) => {
        dump(routeMap.get(childId)!, indent + '  ')
      })
      output += `${indent}</Route>\n`
    }
  }
  let output = '<Routes>\n'
  dump(routeMap.get('root')!, '  ')
  output += '</Routes>\n'
  console.log(output)
}
