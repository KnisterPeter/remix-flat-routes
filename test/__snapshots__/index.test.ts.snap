// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`define routes should allow routes to specify different parent routes 1`] = `
Object {
  "routes/parent": Object {
    "caseSensitive": undefined,
    "file": "routes/parent.tsx",
    "id": "routes/parent",
    "index": undefined,
    "parentId": "root",
    "path": "parent",
  },
  "routes/parent.some.nested": Object {
    "caseSensitive": undefined,
    "file": "routes/parent.some.nested.tsx",
    "id": "routes/parent.some.nested",
    "index": undefined,
    "parentId": "routes/parent",
    "path": "some/nested",
  },
  "routes/parent.some_.nested.route": Object {
    "caseSensitive": undefined,
    "file": "routes/parent.some_.nested.route.tsx",
    "id": "routes/parent.some_.nested.route",
    "index": undefined,
    "parentId": "routes/parent",
    "path": "some/nested/route",
  },
}
`;

exports[`define routes should correctly nest routes 1`] = `
Object {
  "routes/app.$organizationSlug": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.tsx",
    "id": "routes/app.$organizationSlug",
    "index": undefined,
    "parentId": "root",
    "path": "app/:organizationSlug",
  },
  "routes/app.$organizationSlug.edit": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.edit.tsx",
    "id": "routes/app.$organizationSlug.edit",
    "index": undefined,
    "parentId": "routes/app.$organizationSlug",
    "path": "edit",
  },
  "routes/app.$organizationSlug.projects": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.projects.tsx",
    "id": "routes/app.$organizationSlug.projects",
    "index": undefined,
    "parentId": "routes/app.$organizationSlug",
    "path": "projects",
  },
  "routes/app.$organizationSlug.projects.$projectId": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.projects.$projectId.tsx",
    "id": "routes/app.$organizationSlug.projects.$projectId",
    "index": undefined,
    "parentId": "routes/app.$organizationSlug.projects",
    "path": ":projectId",
  },
  "routes/app.$organizationSlug.projects.$projectId.edit": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.projects.$projectId.edit.tsx",
    "id": "routes/app.$organizationSlug.projects.$projectId.edit",
    "index": undefined,
    "parentId": "routes/app.$organizationSlug.projects.$projectId",
    "path": "edit",
  },
  "routes/app.$organizationSlug.projects.new": Object {
    "caseSensitive": undefined,
    "file": "routes/app.$organizationSlug.projects.new.tsx",
    "id": "routes/app.$organizationSlug.projects.new",
    "index": undefined,
    "parentId": "routes/app.$organizationSlug.projects",
    "path": "new",
  },
}
`;

exports[`define routes should define routes for complex structure 1`] = `
Object {
  "routes/_auth": Object {
    "caseSensitive": undefined,
    "file": "routes/_auth.tsx",
    "id": "routes/_auth",
    "index": undefined,
    "parentId": "root",
    "path": undefined,
  },
  "routes/_auth.forgot-password": Object {
    "caseSensitive": undefined,
    "file": "routes/_auth.forgot-password.tsx",
    "id": "routes/_auth.forgot-password",
    "index": undefined,
    "parentId": "routes/_auth",
    "path": "forgot-password",
  },
  "routes/_auth.login": Object {
    "caseSensitive": undefined,
    "file": "routes/_auth.login.tsx",
    "id": "routes/_auth.login",
    "index": undefined,
    "parentId": "routes/_auth",
    "path": "login",
  },
  "routes/_auth.reset-password": Object {
    "caseSensitive": undefined,
    "file": "routes/_auth.reset-password.tsx",
    "id": "routes/_auth.reset-password",
    "index": undefined,
    "parentId": "routes/_auth",
    "path": "reset-password",
  },
  "routes/_auth.signup": Object {
    "caseSensitive": undefined,
    "file": "routes/_auth.signup.tsx",
    "id": "routes/_auth.signup",
    "index": undefined,
    "parentId": "routes/_auth",
    "path": "signup",
  },
  "routes/_landing": Object {
    "caseSensitive": undefined,
    "file": "routes/_landing.tsx",
    "id": "routes/_landing",
    "index": undefined,
    "parentId": "root",
    "path": undefined,
  },
  "routes/_landing.about": Object {
    "caseSensitive": undefined,
    "file": "routes/_landing.about.tsx",
    "id": "routes/_landing.about",
    "index": undefined,
    "parentId": "routes/_landing",
    "path": "about",
  },
  "routes/_landing.index": Object {
    "caseSensitive": undefined,
    "file": "routes/_landing.index.tsx",
    "id": "routes/_landing.index",
    "index": true,
    "parentId": "routes/_landing",
    "path": undefined,
  },
  "routes/app": Object {
    "caseSensitive": undefined,
    "file": "routes/app.tsx",
    "id": "routes/app",
    "index": undefined,
    "parentId": "root",
    "path": "app",
  },
  "routes/app.calendar": Object {
    "caseSensitive": undefined,
    "file": "routes/app.calendar.tsx",
    "id": "routes/app.calendar",
    "index": undefined,
    "parentId": "routes/app",
    "path": "calendar",
  },
  "routes/app.calendar.$day": Object {
    "caseSensitive": undefined,
    "file": "routes/app.calendar.$day.tsx",
    "id": "routes/app.calendar.$day",
    "index": undefined,
    "parentId": "routes/app.calendar",
    "path": ":day",
  },
  "routes/app.calendar.index": Object {
    "caseSensitive": undefined,
    "file": "routes/app.calendar.index.tsx",
    "id": "routes/app.calendar.index",
    "index": true,
    "parentId": "routes/app.calendar",
    "path": undefined,
  },
  "routes/app.projects": Object {
    "caseSensitive": undefined,
    "file": "routes/app.projects.tsx",
    "id": "routes/app.projects",
    "index": undefined,
    "parentId": "routes/app",
    "path": "projects",
  },
  "routes/app.projects.$id": Object {
    "caseSensitive": undefined,
    "file": "routes/app.projects.$id.tsx",
    "id": "routes/app.projects.$id",
    "index": undefined,
    "parentId": "routes/app.projects",
    "path": ":id",
  },
  "routes/app_.projects.$id.roadmap": Object {
    "caseSensitive": undefined,
    "file": "routes/app_.projects.$id.roadmap.tsx",
    "id": "routes/app_.projects.$id.roadmap",
    "index": undefined,
    "parentId": "root",
    "path": "app/projects/:id/roadmap",
  },
  "routes/app_.projects.$id.roadmap[.pdf]": Object {
    "caseSensitive": undefined,
    "file": "routes/app_.projects.$id.roadmap[.pdf].tsx",
    "id": "routes/app_.projects.$id.roadmap[.pdf]",
    "index": undefined,
    "parentId": "root",
    "path": "app/projects/:id/roadmap.pdf",
  },
}
`;

exports[`define routes should define routes for flat-files 1`] = `
Object {
  "routes/$lang.$ref": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref.tsx",
    "id": "routes/$lang.$ref",
    "index": undefined,
    "parentId": "root",
    "path": ":lang/:ref",
  },
  "routes/$lang.$ref.$": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref.$.tsx",
    "id": "routes/$lang.$ref.$",
    "index": undefined,
    "parentId": "routes/$lang.$ref",
    "path": "*",
  },
  "routes/$lang.$ref._index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref._index.tsx",
    "id": "routes/$lang.$ref._index",
    "index": true,
    "parentId": "routes/$lang.$ref",
    "path": undefined,
  },
  "routes/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/_index.tsx",
    "id": "routes/_index",
    "index": true,
    "parentId": "root",
    "path": undefined,
  },
  "routes/healthcheck": Object {
    "caseSensitive": undefined,
    "file": "routes/healthcheck.tsx",
    "id": "routes/healthcheck",
    "index": undefined,
    "parentId": "root",
    "path": "healthcheck",
  },
}
`;

exports[`define routes should define routes for flat-folders 1`] = `
Object {
  "routes/$lang.$ref.$/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref.$/_index.tsx",
    "id": "routes/$lang.$ref.$/_index",
    "index": undefined,
    "parentId": "routes/$lang.$ref/_index",
    "path": "*",
  },
  "routes/$lang.$ref._index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref._index/_index.tsx",
    "id": "routes/$lang.$ref._index/_index",
    "index": true,
    "parentId": "routes/$lang.$ref/_index",
    "path": undefined,
  },
  "routes/$lang.$ref/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref/_index.tsx",
    "id": "routes/$lang.$ref/_index",
    "index": undefined,
    "parentId": "root",
    "path": ":lang/:ref",
  },
  "routes/_index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/_index/_index.tsx",
    "id": "routes/_index/_index",
    "index": true,
    "parentId": "root",
    "path": undefined,
  },
  "routes/healthcheck/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/healthcheck/_index.tsx",
    "id": "routes/healthcheck/_index",
    "index": undefined,
    "parentId": "root",
    "path": "healthcheck",
  },
}
`;

exports[`define routes should define routes for flat-folders on Windows 1`] = `
Object {
  "routes/$lang.$ref.$/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref.$\\\\_index.tsx",
    "id": "routes/$lang.$ref.$/_index",
    "index": undefined,
    "parentId": "routes/$lang.$ref/_index",
    "path": "*",
  },
  "routes/$lang.$ref._index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref._index\\\\_index.tsx",
    "id": "routes/$lang.$ref._index/_index",
    "index": true,
    "parentId": "routes/$lang.$ref/_index",
    "path": undefined,
  },
  "routes/$lang.$ref/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref\\\\_index.tsx",
    "id": "routes/$lang.$ref/_index",
    "index": undefined,
    "parentId": "root",
    "path": ":lang/:ref",
  },
  "routes/_index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/_index\\\\_index.tsx",
    "id": "routes/_index/_index",
    "index": true,
    "parentId": "root",
    "path": undefined,
  },
  "routes/healthcheck/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/healthcheck\\\\_index.tsx",
    "id": "routes/healthcheck/_index",
    "index": undefined,
    "parentId": "root",
    "path": "healthcheck",
  },
}
`;

exports[`define routes should ignore non-route files in flat-folders 1`] = `
Object {
  "routes/$lang.$ref._index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref._index/_index.tsx",
    "id": "routes/$lang.$ref._index/_index",
    "index": true,
    "parentId": "routes/$lang.$ref/_layout",
    "path": undefined,
  },
  "routes/$lang.$ref/_layout": Object {
    "caseSensitive": undefined,
    "file": "routes/$lang.$ref/_layout.tsx",
    "id": "routes/$lang.$ref/_layout",
    "index": undefined,
    "parentId": "root",
    "path": ":lang/:ref",
  },
  "routes/_index/_index": Object {
    "caseSensitive": undefined,
    "file": "routes/_index/_index.tsx",
    "id": "routes/_index/_index",
    "index": true,
    "parentId": "root",
    "path": undefined,
  },
  "routes/healthcheck/_route": Object {
    "caseSensitive": undefined,
    "file": "routes/healthcheck/_route.tsx",
    "id": "routes/healthcheck/_route",
    "index": undefined,
    "parentId": "root",
    "path": "healthcheck",
  },
}
`;

exports[`define routes should support markdown routes as flat-files 1`] = `
Object {
  "routes/docs": Object {
    "caseSensitive": undefined,
    "file": "routes/docs.tsx",
    "id": "routes/docs",
    "index": undefined,
    "parentId": "root",
    "path": "docs",
  },
  "routes/docs.readme": Object {
    "caseSensitive": undefined,
    "file": "routes/docs.readme.md",
    "id": "routes/docs.readme",
    "index": undefined,
    "parentId": "routes/docs",
    "path": "readme",
  },
}
`;

exports[`define routes should support markdown routes as flat-folders 1`] = `
Object {
  "routes/docs/_layout": Object {
    "caseSensitive": undefined,
    "file": "routes/docs/_layout.tsx",
    "id": "routes/docs/_layout",
    "index": undefined,
    "parentId": "root",
    "path": "docs",
  },
  "routes/docs/readme.route": Object {
    "caseSensitive": undefined,
    "file": "routes/docs/readme.route.mdx",
    "id": "routes/docs/readme.route",
    "index": undefined,
    "parentId": "routes/docs/_layout",
    "path": "readme",
  },
}
`;
